taxcalc/utils.py

```diff
@@ -197,14 +196,19 @@ def groupby_income_bins(df, bins=None, right=True):
             indicate (1,2], (2,3], (3,4].
 
     """
-
-    if not bins:
+    if compare_with == "tpc":
         bins = [-1e14, 0, 9999, 19999, 29999, 39999, 49999, 74999, 99999,
                 200000, 1e14]
+    elif compare_with == "soi":
+        bins = [-1e14, 0, 4999, 9999, 14999, 19999, 24999, 29999, 39999, 49999, 74999, 99999,
+                199999, 499999, 999999, 1499999,1999999, 4999999, 9999999, 1e14]
+    elif not bins:
```

Comment:
For this function, if `compare_with` is not specified, you will get the bins of line 203 regardless of whether you specify `bins` or not. If `compare_with` == `tpc`, you will get the bins of line 205 regardless of whether you specify `bins` or not. The only way to specify your own `bins` is to pass `compare_with` as something other than `soi` or `tpc`. I think this will be error prone. In general, I think that by not specifying `compare_with` and specifying your own `bins`, you should get those bins.